x-pack/filebeat/input/entityanalytics/internal/kvstore/store.go

```diff
@@ -0,0 +1,99 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kvstore
+
+import (
+	"fmt"
+	"os"
+
+	"go.etcd.io/bbolt"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type TxDoneFunc func(tx *Transaction, err error) error
+
+// Store is a key/value store with transaction capabilities.
+type Store struct {
+	db     *bbolt.DB
+	logger *logp.Logger
+}
+
+// RunTransaction runs a transaction. Multiple read-only transactions may be
+// started, but only one write transaction is allowed at any given time.
+//
+// To close out a transaction, either commit the changes or rollback. For
+// read-only transactions, no changes can be committed, so calling commit or
+// rollback results in the transaction being "rolled back".
+func (s *Store) RunTransaction(writable bool, fn func(tx *Transaction) error) (err error) {
+	var t Transaction
+
+	t.writeable = writable
+	t.tx, err = s.db.Begin(writable)
+	if err != nil {
+		return fmt.Errorf("unable to begin transaction: %w", err)
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
What panics are expected to be recovered here?

Reply:
If something panics in the transaction function (in `fn func(tx *Transaction) error`), we want to make sure the transaction lock is released from database.

Reply:
OK. Can we note in the logs that that should be reported? I feel queasy about recovering to good state when there are locks involved and would generally prefer to crash in that case, but in other places we do the recover.

Reply:
I think as long as we tumble out completely and it doesn't leave the database file in some weird inconsistent state (which I don't think it should), then it should be fine. I'll run a couple tests to make sure that's the case.